feat(router): Add router configuration to resolve navigation promise on error

atscott opened the pull request:
With the deprecation of the configurable errorHandler in the Router, there is a missing
use-case to prevent the navigation promise from rejecting on an error. This rejection
results in unhandled promise rejections. This commit allows developers to instruct
the router to instead resolve the navigation promise with 'false', which matches
the behavior of other failed navigations.

Resolving the Promise would be the ideal default behavior. It is rare
that any code handles the navigation Promise at all and even more rare
that the Promise rejection is caught.

fixes https://github.com/angular/angular/issues/48902

alfaproject commented on packages/router/test/integration.spec.ts:
```diff
@@ -2332,8 +2332,7 @@ describe('Integration', () => {
       try {
         await router.navigateByUrl('/throwing');
         fail('navigation should throw');
```
i guess it shouldn't throw anymore, right?

atscott committed:
feat(router): Add router configuration to resolve navigation promise on error

With the deprecation of the configurable errorHandler in the Router, there is a missing
use-case to prevent the navigation promise from rejecting on an error. This rejection
results in unhandled promise rejections. This commit allows developers to instruct
the router to instead resolve the navigation promise with 'false', which matches
the behavior of other failed navigations.

Resolving the Promise would be the ideal default behavior. It is rare
that any code handles the navigation Promise at all and even more rare
that the Promise rejection is caught. Updating the default value for
this option should be considered for an upcoming major version.

fixes #48902

atscott commented:
Updated approach to be an opt-in rather than changing the default behavior of the error handler. This proved to be difficult and time-consuming to migrate.

AndrewKushnir commented on packages/router/src/navigation_transition.ts:
```diff
@@ -718,7 +718,20 @@ export class NavigationTransitions {
                                overallTransitionState.resolve(router.errorHandler(e));
                              } catch (ee) {
-                               overallTransitionState.reject(ee);
+                               // TODO(atscott): consider flipping the default behavior of
+                               // resolveNavigationPromiseOnError to be `resolve(false)` when
+                               // undefined. This is the most sane thing to do given that
+                               // applications very rarely handle the promise rejection and, as a
+                               // result, would get "unhandled promise rejection" console logs.
```
+1

dylhunn reviewed (approved):
reviewed-for: public-api

alan-agius4 reviewed (approved):
LGTM

jessicajaniuk reviewed (approved):
reviewed-for: public-api

dylhunn commented:
This PR was merged into the repository by commit 50d79162785bb8d3e158a7a4a3733f4c75d3b127.
